mediacloud/mediawords/dbi/downloads.py

```diff
@@ -0,0 +1,388 @@
+    """Return the store from which to read the content for the given download."""
+    download_store = None
+
+    config = mediawords.util.config.get_config()
+
+    if config['mediawords'].get('read_all_downloads_from_s3', False):
+        return _get_amazon_s3_store()
+
+    path = download.get('path', 's3:')
+
+    match = re.search('^([\w]+):', path)
+    location = match.group(1) if match else 's3'
+    location = location.lower()
+
+    if location == 'content':
+        download_store = _get_inline_store()
+    elif location == 'postgresql':
+        download_store = _get_postgresql_store()
+    elif location in ('s3', 'amazon_s3'):
+        download_store = _get_amazon_s3_store()
+    elif location == 'gridfs' or location == 'tar':
+        # these are old storage formats that we moved to postgresql
+        download_store = _get_postgresql_store()
+    else:
+        id = download.get('downloads_id', '(no downloads_id')
+        raise McDBIDownloadsException("Location 'location' is unknown for download %d", [id])
+
+    assert download_store is not None
+
+    return download_store
+
+
+def fetch_content(db: DatabaseHandler, download: dict) -> str:
+    """Fetch the content for the given download from the configured content store."""
+    if 'downloads_id' not in download:
+        raise McDBIDownloadsException("downloads_id not in download")
+
+    if not download_successful(download):
+        raise McDBIDownloadsException(
+            "attempt to fetch content for unsuccessful download: %d" % (download['downloads_id']))
+
+    store = _get_store_for_reading(download)
+
+    content_bytes = store.fetch_content(db, download['downloads_id'], download['path'])
+
+    content = content_bytes.decode()
+
+    # horrible hack to fix old content that is not stored in unicode
+    config = mediawords.util.config.get_config()
+    ascii_hack_downloads_id = config['mediawords'].get('ascii_hack_downloads_id', 0)
+    if download['downloads_id'] < ascii_hack_downloads_id:
+        # this matches all non-printable-ascii characters.  python re does not support POSIX character
+        # classes like [[:ascii:]]
+        content = re.sub(r'[^ -~]', ' ', content)
+
+    return content
+
+
+def store_content(db: DatabaseHandler, download: dict, content: str) -> dict:
+    """Store the content for the download."""
+    # feed_error state indicates that the download was successfull but that there was a problem
+    # parsing the feed afterward.  so we want to keep the feed_error state even if we redownload
+    # the content
+    new_state = 'success' if download['state'] != 'feed_error' else 'feed_error'
+
+    try:
+        path = _get_store_for_writing().store_content(db, download['downloads_id'], content)
+        error = ''
+    except Exception as e:
+        raise McDBIDownloadsException("error while trying to store download %d: %s" % (download['downloads_id'], e))
+        new_state = 'error'
```

Comment:
<img width="839" alt="screenshot 2018-02-20 05 23 28" src="https://user-images.githubusercontent.com/181949/36397558-426cae40-15fe-11e8-8cb4-04052b14110a.png">

Row in `downloads` will never get its `state` set to `error`.